algorithms/HPdutility/R/hpdsample.R

```diff
@@ -0,0 +1,301 @@
+# as published by the Free Software Foundation; either version 2                    #
+# of the License, or (at your option) any later version.                            #
+#                                                                                   #
+# This program is distributed in the hope that it will be useful,                   #
+# but WITHOUT ANY WARRANTY; without even the implied warranty of                    #
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                      #
+# GNU General Public License for more details.                                      #
+#                                                                                   #
+# You should have received a copy of the GNU General Public License                 #
+# along with this program; if not, write to the Free Software                       #
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.    #
+#####################################################################################
+
+
+hpdsample <- function(data1, data2, nSampParts, sampRatio, trace = FALSE) {
+  # Samples data from data1 (and when present, data2, maintaining the same
+  # order of samples between them) into a new darray/dframe with nSampParts
+  # partitions, each of which contains approximately sampRatio * nrow(data1)
+  # elements. 
+  #
+  # Args:
+  #   data1:            A darray/dframe to sample
+  #   data2 (optional): A darray/dframe to sample along with data1. Must have
+  #                     the same number of rows and partitions as data1. Each
+  #                     row of data2 is assumed to correspond with the same row
+  #                     of data1, and this correspondence will be maintained
+  #                     when sampling.
+  #   nSampParts:       The number of output partitions for the sample data
+  #   sampRatio:        The percentage of data from data1/data2 to be sampled
+  #                     in each output partition. E.g. if nrow(data1) = 100,
+  #                     and sampRatio is 0.1, each output partition will
+  #                     contain approximately 100 * 0.1 = 10 rows.
+  #   trace:            A boolean that states whether or not to print the
+  #                     progress of the distributed foreach calls
+  # Returns:
+  #   If only data1 is present, an output dobject of the same type as data1
+  #   with nSampParts partitions, each of which contains sampRatio *
+  #   nrow(data1) elements sampled randomly from data1. If data2 is present, a
+  #   list with elements sdata1 (containing the data sampled from data1) and
+  #   sdata2 (containing the data sampled from data2)
+
+  if (!is.numeric(nSampParts) || length(nSampParts) != 1 || nSampParts <= 0 ||
+      nSampParts %% 1 != 0) 
+    stop("'nSampParts' must be a positive integer")
+
+  if (!is.numeric(sampRatio) || length(sampRatio) != 1 || sampRatio <= 0) 
```

Comment:
I will feel better if we limit the max value of sampRatio; e.g., 1. Unlimited value cannot be tested, and can eventually break the code somewhere.